### ISSUE TITLE
MacOS startup is slow on the app's first launch after install

### ISSUE THREAD

**Issue by emlys** [natcap/invest-workbench#158]
Opening the workbench (`npm run start`) is sometimes very slow. When I tried it on `main` just now, there was a blank splash screen for nearly a full minute. Then the InVEST logo appeared on the splash screen for a few seconds, and then the app opened. The splash screen appears in the center of the screen and cannot be moved, and it always appears on top of other windows. This almost prevents you from doing other things while waiting for it to load.

I'm not sure what electron is doing that takes so much time before even loading the splash screen image, but if it's not avoidable, we should keep the splash screen from being annoying to the user.

**Comment by davemfish** [comment]
Oh interesting. I'd be very interested to figure out why startup is so slow and focus on that problem first. It's also worth checking the startup time with an installed application - that tends to always be faster than running `npm run start`.

**Comment by davemfish** [comment]
We can change `alwaysOnTop: true` to `false` in the splash screen BrowserWindow instance.

**Comment by phargogh** [comment]
When I tried to launch the workbench with the two-shell dev setup we talked about this morning, the splash screen wasn't up for more than a couple seconds, which seemed pretty reasonable to me.  I did launch it yesterday and the application took longer to start up, but it all seemed like a reasonable startup time.  Sorry I don't have any other info at the moment!

**Comment by emlys** [comment]
I couldn't reproduce this again exactly. With a fresh clone of the workbench, the first `npm start` was still very slow:
- No window appears, cursor has loading icon, for ~30 seconds
- natcap logo splash screen appears for ~20 seconds (didn't see a blank splash screen this time)
- Blank app window appears behind the splash screen for ~5 seconds
- App window finishes loading but the splash screen stays there forever.

Subsequently `npm start`s took just a second or two and the splash screen behaved as expected.

Probably unrelated, but `npm start` gives me a warning I don't remember seeing before:
```
(node:3046) ExtensionLoadWarning: Warnings loading extension at /Users/emily/Library/Application Support/invest-workbench/extensions/fmkadmapgofadopljbjfkapdkoienihi:
  Unrecognized manifest key 'browser_action'.
  Unrecognized manifest key 'minimum_chrome_version'.
  Unrecognized manifest key 'update_url'.
  Cannot load extension with file or directory name _metadata. Filenames starting with "_" are reserved for use by the system.

(Use `Electron --trace-warnings ...` to show where the warning was created)
```

**Comment by emlys** [comment]
Opening a freshly downloaded copy of the workbench app for the first time:
- Right click > Open
- Invest icon appears in the bar and bounces indicating loading for ~55 seconds
- Initial unverified developer warning appears. There is no option to open anyway, only "Move to Trash" or "Cancel"
- Click "Cancel". Right click and open the app again.
- This time I immediately get the next unverified developer warning. This one has the option to "Open".
- Click "Open".
- Nothing apparently happens for a few seconds (I would think the icon should bounce again here)
- Splash screen appears briefly over the empty app window
- App loads as expected

**Comment by emlys** [comment]
So the slow loading is still a problem for the app, and the user experience starting it for the first time isn't great. But maybe the splash screen problem is just something we'll encounter occasionally in development.

**Comment by davemfish** [comment]
Well this is all very interesting. Thanks for the rundowns! It does sound like there is some slow electron startup that happens only the first time the app is launched. Not ideal, but maybe we can look into that. There's very little code in our `main.js` that runs before the splash screen loads, so the ~50s-ish must be something electron always does and maybe others have figured out that problem.

> Initial unverified developer warning appears. There is no option to open anyway, only "Move to Trash" or "Cancel"
Click "Cancel". Right click and open the app again.
This time I immediately get the next unverified developer warning. This one has the option to "Open".
Click "Open".

This sounds like another isolated problem to look into. I can't recall the state of our code-signing capabilities right now. But we have #28 for this.

**Comment by davemfish** [comment]
```
(node:3046) ExtensionLoadWarning: Warnings loading extension at /Users/emily/Library/Application Support/invest-workbench/extensions/fmkadmapgofadopljbjfkapdkoienihi:
  Unrecognized manifest key 'browser_action'.
  Unrecognized manifest key 'minimum_chrome_version'.
  Unrecognized manifest key 'update_url'.
  Cannot load extension with file or directory name _metadata. Filenames starting with "_" are reserved for use by the system.

(Use `Electron --trace-warnings ...` to show where the warning was created)
```
These warnings are related to loading the react-devtools extension, which was a recent change. There's been a lot of discussion of them and my conclusion was that it's simply not worth the effort to try to silence them. Or rather, the bug is upstream and silencing them would require each of us to jump through some manual hoops in our local filesystem, which I never bothered trying. I should have warned you all though!

https://github.com/MarshallOfSound/electron-devtools-installer/issues/143
https://stackoverflow.com/questions/23918539/extension-installed-from-chrome-webstore-throws-error-package-is-invalid-detai

**Comment by davemfish** [comment]
I kinda hope the slow startup does not come down to doing #151

**Comment by davemfish** [comment]
I've recently tested this on Mac as well. I don't have any problems with slow startup during development. When I install a production build, I have a similar experience to Emily:

- Right click > Open
- Invest icon appears in the bar and bounces indicating loading for ~55 seconds
- Initial unverified developer warning appears. There is no option to open anyway, only "Move to Trash" or "Cancel"
- Click "Cancel". Right click and open the app again.
- This time I immediately get the next unverified developer warning. This one has the option to "Open".
- Click "Open".
- Nothing apparently happens for a few seconds (I would think the icon should bounce again here)
- Splash screen appears briefly over the empty app window
- App loads as expected

It's only slow the first time I launch the app. If I quit and launch again, it responds right away and loads within a few seconds.

**Comment by davemfish** [comment]
I'm not exactly sure where to start with debugging this. One idea is to first deal with codesigning for MacOS and then see if this is still a problem. #28

**Comment by davemfish** [comment]
I renamed this issue because the splash screen part of it is resolved, so it's really just about startup time on MacOS now.

**Comment by davemfish** [comment]
I'm not sure if it's related to the same underlying issue here or not, but the puppeteer test on MacOS requires much longer time than on Windows (2m30s vs 19s) and a lot of it seems to be waiting for the electron app to startup before puppeteer is allowed to connect.

The very long jest timeouts in that script are there to accommodate this. E.g:

```javascript
await waitFor(() => {
  expect(BROWSER && BROWSER.isConnected()).toBeTruthy();
}, { timeout: 60000 });
```

**Comment by davemfish** [comment]
Based on some recent testing by me and @phargogh , after installing and opening the app, we both experience two periods of waiting. 

First wait: ~30s while the OS security is "Verifying" the exe. On older versions of macos (High Sierra) there is a progress bar during this. On more recent versions there is no indicator that anything is happening.

Second wait: After verifying, the user can Open anyway, which starts another ~30-60s wait where there is no status indicator at all. 

Both periods seem to be waiting on macos Gatekeeper or other security processes that are out of our control (see Slack thread for more details). The exact programs running vary by OS version. It's not clear whether a fully notarized app would speed any of this up.

This all seems out of our control, unfortunately. @phargogh do you think we just mark this as wontfix at this point?

**Comment by davemfish** [comment]
This seems to be mostly out of our hands. We have other issues linked above that we can work on to possibly alleviate this problem. And an issue for codesigning, which there's reason to think might help. #28 . I'm going to close this one in favor of those other open issues.
